include/zephyr/drivers/pwm/max31790.h

```diff
@@ -0,0 +1,60 @@
+/*
```

Comment:
This is not addressed, this header is used only in the driver, but is available to the public. Application using this header is not portable/generic, therefore this header can be placed in drivers/pwm/. Driver specific application can include it from there using zephyr_include_directories() and there is no need to pollute public header directory.

Reply:
I just checked and there are plenty of examples of driver specific headers in include/zephyr/drivers/*:

- maxim_ds3231.h
- ccs811.h
- fdc2x1x.h
- hl7800.h

Therefore I assume this is the way to go?

Reply:
> I just checked and there are plenty of examples of driver specific headers in include/zephyr/drivers/*:
> 
>     * maxim_ds3231.h
> 
>     * ccs811.h
> 
>     * fdc2x1x.h
> 
>     * hl7800.h
> 

maxim_ds3231.h, ccs811.h, and hl7800.h provide a driver specific APIs (and there is an exception for that).
fdc2x1x.h  should not be there.

> 
> Therefore I assume this is the way to go?

Aside from the fact that it is not necessary, if something changes tomorrow or the day after, e.g. API or other changes to the headers directory structure, how would anyone know what it is there for? Not even the git log would help, because there is no justification at all in the commit message.

Reply:
Let us focus for a moment on the definition of a driver specific API, because I don't yet understand the difference here. In the case of the PWM drivers there are the upper 8 bits of pwm_flags_t explicitly reserved for SoC specific flags. I would consider the usage of these bits as the usage of a driver specific API. Therefore, I think the exposure of how these flags are interpreted for this specific driver is desirable. Or do I understand this incorrectly?

Secondly, regarding the documentation. A simple grep with one of these defines shows that they are used within this driver for the SoC specific flags. Is this insufficient?

Reply:
`fdc2x1x.h` exposes some device specific internal attribute, that's the way to do it, its own registers are in the "private" include.

Since they are custom flags though I think this should go in `include/zephyr/dt-bindings/pwm`, like https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dt-bindings/pwm/stm32_pwm.h

Reply:
I really think it makes sense to expose these things via header, which is automatically available. And there are plenty of examples where similar stuff has already been done with other drivers. We can discuss which is the correct directory, as I see your point that these are not used within the devicetree.

At the moment I don't get why this specific driver header should be different than the others and require a user to manually include it via zephyr_include_directories. But I'll happily change it if there are good reasons to do so.

Reply:
Sorry I feel like I'm adding confusion now --- how are those constant meant to be used?

Reply:
> At the moment I don't get why this specific driver header should be different than the others and require a user to manually include it via zephyr_include_directories. But I'll happily change it if there are good reasons to do so.

I do not understand how one can justify wrong behavior with something someone did not do right before.
This driver is using generic PWM driver API and is a PWM controller, right? Then application should use generic PWM driver API to interface with the controller.

Reply:
For instance as argument of pwm_set_cycles, see:
https://docs.zephyrproject.org/apidoc/latest/group__pwm__interface.html#gaff280789f7b45fdefc354b3f841fe3ef
https://docs.zephyrproject.org/apidoc/latest/group__pwm__interface.html#gae1dcfb878163da76041efcedf0960fa0

Reply:
> > At the moment I don't get why this specific driver header should be different than the others and require a user to manually include it via zephyr_include_directories. But I'll happily change it if there are good reasons to do so.
> 
> I do not understand how one can justify wrong behavior with something someone did not do right before. This driver is using generic PWM driver API and is a PWM controller, right? Then application should use generic PWM driver API to interface with the controller.

First of all, I consider consistency rather important. If all these other occasions are wrong, shouldn't we fix them?
Second, these flags are part of the generic PWM driver API, as described in the documentation (see the links above).

Reply:
> First of all, I consider consistency rather important. If all these other occasions are wrong, shouldn't we fix them?

I do not see any PWM driver specific includes in the tree. Consistency is a generic approach for all drivers.

> Second, these flags are part of the generic PWM driver API, as described in the documentation (see the links above).

I see "The upper 8 bits are reserved for SoC specific flags." , but is it intended to be used with external PWM controllers?

https://github.com/zephyrproject-rtos/zephyr/blob/e5fcca6e99a52334c0c726dfc1b0c432f8d944bb/include/zephyr/drivers/pwm.h#L70-L79

Also from your header "Configures the internal control loop of the fan." does not sound like PWM configuration but specific fan controller setting. Is it abusing the use of pwm_flags_t?

Reply:
I look at this similarly to an ADC integrated in a SoC and an external one, for instance connected via SPI. These devices are usually handled similarly in zephyr, therefore I would also not make a difference here.

Of course, there is a rather profound problem: In which area should a driver for this hardware get implemented? It is on one hand a PWM IC, but on the other hand on top of that a fan controller as well. My approach was to implement both functionalities as a PWM driver, as PWM allows this driver specific extension.

Reply:
> Of course, there is a rather profound problem: In which area should a driver for this hardware get implemented? It is on one hand a PWM IC, but on the other hand on top of that a fan controller as well. My approach was to implement both functionalities as a PWM driver, as PWM allows this driver specific extension.

Maybe hwmon, but I would not do it yet.
https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/max31790.c

The commit message should clearly describe what type of controller this driver is for. If misuse of pwm_flags_t is acceptable, the commit message should include a justification for why the specific header is needed and how it must be used. As I understand it, this controller cannot be used by generic PWM API only, and always requires specific flags, right?

Reply:
Making the commit message more expressive is no problem, already done and pushed.

The driver defaults to using the controller in PWM mode. Only when the RPM_MODE flag is set the controller is configured as fan controller.